mlir/lib/Dialect/Linalg/Utils/Utils.cpp

```diff
@@ -870,5 +870,80 @@ getReassociationMapForFoldingUnitDims(ArrayRef<OpFoldResult> mixedSizes) {
   return reassociation;
 }
 
+// Returns true if the value is a constant float or integer.
+bool isValConstZero(Value val) {
+  return matchPattern(val, m_AnyZeroFloat()) || matchPattern(val, m_Zero());
+}
+
+// Returns true if the attribute represent "all zeros".
+static bool isZeroAttr(Attribute attribute) {
+  return TypeSwitch<Attribute, bool>(attribute)
+      .Case<FloatAttr>([](auto attr) { return attr.getValueAsDouble() == 0.0; })
+      .Case<IntegerAttr>([](auto attr) { return attr.getInt() == 0; })
+      .Case<DenseElementsAttr>([](auto attr) {
+        if (!attr.getElementType().isIntOrFloat())
+          return false;
+        if (!attr.isSplat())
+          return false;
+        auto splat = attr.template getSplatValue<Attribute>();
+        return isZeroAttr(splat);
```

Comment:
Why do you need to recurse into `isZeroAttr` in this case? Also:
```suggestion
        // Splat of 0
        if (attr.isSplat() && (attr.getSplatValue<APInt>().isZero() || attr.getSplatValue<APFloat>().isZero())
          return true;
```

Hope I didn't miss anything myself.

Reply:
Turns out the "predicate matcher" function `m_AnyZeroFloat` and `m_Zero` already deal with splatted zeroes.

Have made all this code substantially simpler.